R/assignment_linter.R

```diff
@@ -64,63 +72,113 @@
-      if (allow_right_assign) paste0("//RIGHT_ASSIGN", if (allow_cascading_assign) "" else "[text() = '->']"),
+      "//LEFT_ASSIGN",
+      "//RIGHT_ASSIGN",
+      "//EQ_ASSIGN",
       "//EQ_SUB",
       "//EQ_FORMALS",
-      if (!allow_pipe_assign) "//SPECIAL[text() = '%<>%']"
+      "//SPECIAL[text() = '%<>%']"
     ),
     "[@line1 < following-sibling::expr[1]/@line1]"
   )
 
-  xpath <- paste(collapse = " | ", c(
-    # always block = (NB: the parser differentiates EQ_ASSIGN, EQ_SUB, and EQ_FORMALS)
-    "//EQ_ASSIGN",
+  op_xpath_parts <- c(
+    if (!"=" %in% operator) "//EQ_ASSIGN",
     # -> and ->> are both 'RIGHT_ASSIGN'
-    if (!allow_right_assign) "//RIGHT_ASSIGN" else if (!allow_cascading_assign) "//RIGHT_ASSIGN[text() = '->>']",
+    glue("//RIGHT_ASSIGN[{ xp_text_in_table(setdiff(c('->', '->>'), operator)) }]"),
     # <-, :=, and <<- are all 'LEFT_ASSIGN'; check the text if blocking <<-.
     # NB: := is not linted because of (1) its common usage in rlang/data.table and
     #   (2) it's extremely uncommon as a normal assignment operator
-    if (!allow_cascading_assign) "//LEFT_ASSIGN[text() = '<<-']",
-    if (!allow_trailing) trailing_assign_xpath,
-    if (!allow_pipe_assign) "//SPECIAL[text() = '%<>%']"
-  ))
+    glue("//LEFT_ASSIGN[{ xp_text_in_table(setdiff(c('<-', '<<-'), operator)) }]"),
+    if (!"%<>%" %in% operator) "//SPECIAL[text() = '%<>%']"
+  )
+  if (!is.null(op_xpath_parts)) {
+    # NB: Also used, essentially, in implicit_assignment_linter. Keep in sync.
+    implicit_assignment_xpath <- "
+      [not(parent::expr[
+        preceding-sibling::*[2][self::IF or self::WHILE]
+        or parent::forcond
+        or parent::expr/*[1][self::OP-LEFT-PAREN]
+      ])]
+    "
+    op_xpath <- paste0(op_xpath_parts, implicit_assignment_xpath, collapse = " | ")
+  } else {
+    op_xpath <- NULL
+  }
 
   Linter(linter_level = "expression", function(source_expression) {
     xml <- source_expression$xml_parsed_content
 
-    bad_expr <- xml_find_all(xml, xpath)
-    if (length(bad_expr) == 0L) {
-      return(list())
-    }
+    lints <- NULL
+    if (!is.null(op_xpath)) {
+      op_expr <- xml_find_all(xml, op_xpath)
 
-    operator <- xml_text(bad_expr)
-    lint_message_fmt <- rep("Use <-, not %s, for assignment.", length(operator))
-    lint_message_fmt[operator %in% c("<<-", "->>")] <-
-      "Replace %s by assigning to a specific environment (with assign() or <-) to avoid hard-to-predict behavior."
-    lint_message_fmt[operator == "%<>%"] <-
-      "Avoid the assignment pipe %s; prefer using <- and %%>%% separately."
+      op_text <- xml_text(op_expr)
+      op_lint_message_fmt <- rep(
+        sprintf(
+          "Use %s for assignment, not %%s.",
+          if (length(operator) > 1L) paste("one of", toString(operator)) else operator
+        ),
+        length(op_text)
+      )
+      op_lint_message_fmt[op_text %in% c("<<-", "->>")] <-
```

Comment:
This path will produce a confusing message for `operator = c("<-", "<<-")` when encountering `"->>"`.

Reply:
Good point... I think the sane approach here is to only add this clarification if neither `<<-` nor `->>` is in `operator`, PTAL.